gc/gc-iceberg/src/main/java/org/projectnessie/gc/iceberg/ExpireSnapshotsProcedure.java

```diff
@@ -0,0 +1,435 @@
+            new StructField(
+                OUTPUT_SNAPSHOT_IDS,
+                DataTypes.createArrayType(DataTypes.LongType),
+                true,
+                Metadata.empty()),
+            new StructField(
+                OUTPUT_EXPIRED_DATA_FILES_COUNT, DataTypes.LongType, true, Metadata.empty()),
+            new StructField(
+                OUTPUT_EXPIRED_MANIFEST_LISTS_COUNT, DataTypes.LongType, true, Metadata.empty()),
+            new StructField(
+                OUTPUT_EXPIRED_MANIFESTS_COUNT, DataTypes.LongType, true, Metadata.empty()),
+            new StructField(
+                OUTPUT_EXPIRED_FILES_LIST,
+                DataTypes.createArrayType(DataTypes.StringType),
+                true,
+                Metadata.empty())
+          });
+
+  public ExpireSnapshotsProcedure(TableCatalog currentCatalog) {
+    super(currentCatalog);
+  }
+
+  @Override
+  public ProcedureParameter[] parameters() {
+    return PARAMETERS;
+  }
+
+  @Override
+  public StructType outputType() {
+    return OUTPUT_TYPE;
+  }
+
+  @Override
+  public String description() {
+    return String.format(
+        "Expires the Iceberg snapshots that are collected by '%s' procedure.",
+        IdentifyExpiredSnapshotsProcedure.PROCEDURE_NAME);
+  }
+
+  @Override
+  public InternalRow[] call(InternalRow internalRow) {
+    String expiryBranchName = internalRow.getString(0);
+    String gcCatalogName = internalRow.getString(1);
+    String gcOutputBranchName = internalRow.getString(2);
+    String gcOutputTableIdentifier = internalRow.getString(3);
+    String gcCheckPointTableIdentifier = internalRow.getString(4);
+    Map<String, String> nessieClientConfig = new HashMap<>();
+    MapData map = internalRow.getMap(5);
+    for (int i = 0; i < map.numElements(); i++) {
+      nessieClientConfig.put(
+          map.keyArray().getUTF8String(i).toString(), map.valueArray().getUTF8String(i).toString());
+    }
+    String runId = !internalRow.isNullAt(6) ? internalRow.getString(6) : null;
+    boolean dryRun = !internalRow.isNullAt(7) && internalRow.getBoolean(7);
+
+    IdentifiedResultsRepo identifiedResultsRepo =
+        new IdentifiedResultsRepo(
+            spark(), gcCatalogName, gcOutputBranchName, gcOutputTableIdentifier);
+
+    runId = updateRunId(gcOutputTableIdentifier, runId, identifiedResultsRepo);
+
+    Map<String, List<Long>> expiredSnapshotsPerContentId =
+        getExpiredSnapshotsPerContentId(runId, identifiedResultsRepo);
+
+    Map<String, List<String>> metadataLocationsPerContentId =
+        getMetadataLocationsPerContentId(runId, identifiedResultsRepo);
+
+    List<InternalRow> outputRows = new ArrayList<>();
+    try (NessieApiV1 api = GCUtil.getApi(nessieClientConfig)) {
+      createGCBranch(expiryBranchName, api);
+      // get Nessie catalog
+      Catalog nessieCatalog = GCUtil.loadNessieCatalog(spark(), gcCatalogName, expiryBranchName);
+      String finalRunId = runId;
+      expiredSnapshotsPerContentId
```

Comment:
This will be very slow, because it can only expire one table at a time.

Reply:
Agree, I will make it multi threaded.